torch/distributed/_spmd/iter_graph_module.py

```diff
@@ -0,0 +1,393 @@
+
+    def lint(self) -> None:
+        self.setup_graph.lint()
+        super().lint()
+        self.cleanup_graph.lint()
+
+
+class IterGraphModule(nn.Module):
+    """
+    ``IterGraphModule`` provides the ability to do cross-iteration optimization.
+    Given a ``fx.GraphModule``, main_gm, ``IterGraphModule`` internally
+    duplicate it to 3 copies and redirect the ``forward`` request to a different
+    ``fx.GraphModule`` based on the iteration count. This allows users to do
+    graph optimizations that across iterations (e.g., moving collective wait in
+    the backward to the forward of the next iteration).
+
+    Note that users must call the APIs provided by ``IterGraphModule`` or
+    ``IterGraph`` to rewrite the graph so that ``IterGraphModule`` can keep the
+    data dependency for all 3 graphs.
+    """
+
+    def __init__(self, main_gm: fx.GraphModule) -> None:
+        super().__init__()
+
+        def _copy_gm(src: fx.GraphModule, graph: fx.Graph) -> fx.GraphModule:
+            gm = fx.GraphModule(src, graph)
+            gm.meta = getattr(graph, "meta", {})
+            return gm
+
+        self.setup_gm = _copy_gm(main_gm, copy.deepcopy(main_gm.graph))
+        self.cleanup_gm = _copy_gm(main_gm, copy.deepcopy(main_gm.graph))
+        self.main_gm = _copy_gm(
+            main_gm,
+            IterGraph(main_gm.graph, self.setup_gm.graph, self.cleanup_gm.graph),
+        )
+
+        self._iter = 0
+        self._max_iters = 0
+        self._previous_output: Tuple[Any, ...] = tuple()
+
+    def setup(self, max_iters: int = 0) -> None:
+        """
+        This method is used to tell IterGraphModule the iterations to train so
+        that IterGraphModule knows which iteration is the last one and can do
+        proper cleanup.
+        """
+        if max_iters <= 0:
+            raise ValueError(f"Incorrect max_iters is set, {max_iters}")
+        self._iter = 0
+        self._max_iters = max_iters
+
+    def _run(self, gm: fx.GraphModule, *args, **kwargs) -> Any:
+        if cast(IterGraph, self.main_gm.graph).num_extra_output > 0:
+            # TODO: a general way to support different types of input and output.
+            assert not kwargs, "Has not supported kwargs now."
+            new_args = args + (self._previous_output)
+            output = gm(*new_args, **kwargs)
+            if self._iter < self._max_iters:
+                assert isinstance(
+                    output, tuple
+                ), f"Only support tuple output now. {type(output)}"
+                num_actual_output = (
+                    len(output) - cast(IterGraph, self.main_gm.graph).num_extra_output
+                )
+                assert num_actual_output > 0
+                self._previous_output = output[num_actual_output:]
+                output = output[:num_actual_output]
+                if len(output) == 1:
+                    output = output[0]
+            return output
+        else:
+            return gm(*args, **kwargs)
```

Comment:
why do we not consider self._previous_output in the args for this FW?

Reply:
For this case, this mean all the optimizations do not produce any cross-iteration movement. As a result, no extra/previous output exist. We have to do this as IterGraphModule will be the entry point of all the optimizations of all the graphs but not all graphs may be required to do cross-iteration movement (if algorithm believes no benefit).